src/blt_mad/helpers.go

```diff
@@ -146,26 +134,96 @@ func withinTolerance(a float64, b float64, e float64) bool {
 	}
 }
 
-func withinToleranceFloatSlice(a []float64, b []float64, e float64) bool {
+func WithinToleranceFloatSlice(a []float64, b []float64, e float64) bool {
 	if reflect.DeepEqual(a, b) {
 		return true
 	} else {
 		for i := 0; i < len(a); i++ {
-			if !withinTolerance(a[i], b[i], e) {
+			if !WithinTolerance(a[i], b[i], e) {
 				return false
 			}
 		}
 	}
 	return true
 }
 
-func readTxtToString(filePath string) (string, error) {
-	// Read the entire file into a byte slice
-	content, err := ioutil.ReadFile(filePath)
+func calculatePercentile(numbers []float64, percentile float64) float64 {
+	sort.Float64s(numbers)
+	index := int(percentile / 100 * float64(len(numbers)-1)) //index corresponding to the percentile
+	lower := numbers[index]
+	upper := numbers[index+1]
+	fractionalPart := percentile/100*float64(len(numbers)-1) - float64(index)
+	value := lower + (upper-lower)*fractionalPart
+	return value
+}
+
+func GetValuesLargerThanPercentile(numbers []float64, percentile float64) []float64 {
+	valueAtPercentile := calculatePercentile(numbers, percentile)
+
+	// Get values larger than percentile% of the data
+	var largerValues []float64
+	for _, num := range numbers {
+		if num > valueAtPercentile {
+			largerValues = append(largerValues, num)
+		}
+	}
+
+	return largerValues
+}
+
+func SaveArrayToFile(fileName string, arr []float64) error {
+	file, err := os.Create(fileName)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	for _, value := range arr {
+		_, err := fmt.Fprintf(file, "%f\n", value)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func TxtIntoArrayFloat64(inputFile string) ([]float64, error) {
+	var floats []float64
+
+	// Open the file
+	file, err := os.Open(inputFile)
 	if err != nil {
-		return "", err
+		return nil, err
+	}
+	defer file.Close()
```

Comment:
This `defer` line should go before the previous `if` statement, because if the condition is true, you exit the function without closing the file.